src/game/client/ingameui.cpp

```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+                break;
+            case ACTIONTYPE_SABOTAGE_BUILDING:
+                can_do = g_theActionManager->Can_Sabotage_Building(
+                    draw->Get_Object(), object_to_interact_with, COMMANDSOURCE_PLAYER);
+                break;
+            default:
+                break;
+        }
+
+        if (!can_do) {
+            continue;
+        }
+
+        if (rule == SELECTION_ANY) {
+            return true;
+        }
+
+        qualify++;
+    }
+
+    return rule == SELECTION_ALL && count > 0 && qualify == count;
+}
+
+bool InGameUI::Can_Selected_Objects_Do_Special_Power(CommandButton *command,
+    const Object *object_to_interact_with,
+    Coord3D *position,
+    SelectionRules rule,
+    unsigned int options,
+    Object *o) const
+{
+    const SpecialPowerTemplate *power = command->Get_Special_Power();
+    bool needs_target_pos = (command->Get_Options() & COMMAND_OPTION_NEED_TARGET_POS) != 0;
+    bool needs_target_object = (command->Get_Options() & COMMAND_OPTION_NEED_TARGET_OBJECT) != 0;
+
+    if (needs_target_object && object_to_interact_with == nullptr) {
+        return false;
+    }
+
+    if (needs_target_pos && position == nullptr) {
+        return false;
+    }
+
+    Drawable *draw;
+
+    if (o != nullptr) {
+        draw = o->Get_Drawable();
+    } else {
+        draw = nullptr;
+    }
+
+    std::list<Drawable *> list;
+
+    if (draw != nullptr) {
+        list.push_back(draw);
+    }
+
+    const std::list<Drawable *> *list2;
+
+    if (list.size() != 0) {
+        list2 = &list;
+    } else {
+        list2 = g_theInGameUI->Get_All_Selected_Drawables();
+    }
+
+    int count = 0;
+    int qualify = 0;
+
+    for (auto it = list2->begin(); it != list2->end(); it++) {
+        Drawable *drawable = *it;
+        count++;
+
+        if (needs_target_object || needs_target_pos) {
+            if (needs_target_object) {
+                if (g_theActionManager->Can_Do_Special_Power_At_Object(
+                        drawable->Get_Object(), object_to_interact_with, COMMANDSOURCE_PLAYER, power, options, true)) {
+                    if (rule == SELECTION_ANY) {
+                        return true;
+                    }
+
+                    qualify++;
+                }
+            } else if (needs_target_pos) {
+                if (g_theActionManager->Can_Do_Special_Power_At_Location(drawable->Get_Object(),
+                        position,
+                        COMMANDSOURCE_PLAYER,
+                        power,
+                        object_to_interact_with,
+                        options,
+                        true)) {
+                    if (rule == SELECTION_ANY) {
+                        return true;
+                    }
+
+                    qualify++;
+                }
+            }
+        } else {
+            if (g_theActionManager->Can_Do_Special_Power(draw->Get_Object(), power, COMMANDSOURCE_PLAYER, options, 1)) {

```

Comment:
`drawable`, no ?
Otherwise this may be null.

Reply:
fixed.